Description: add support for dhcp-allowed-srvids

Author: Marek Skrobacki
Origin: other
Last-Update: 2024-05-21 
---
This patch add support for allowing custom Server IDs in the containerised environments.
Built for Debian 12 / dnsmasq 2.89
--- a/man/dnsmasq.8
+++ b/man/dnsmasq.8
@@ -1985,6 +1985,25 @@
 The two-address form of shared-network is also usable with a DHCP relay: the first address
 is the address of the relay and the second, as before, specifies an extra subnet which
 addresses may be allocated from.
+.TP
+.B --dhcp-allowed-srvids[=<ip addr>]
+(IPv4 Only) If the DHCP server is running inside a container or behind a load balancer, the
+DHCPREQUEST arriving at dnsmasq for processing may have a Server ID (option 54)
+configured with an IP address that is not assigned to the local interface. In
+this case, dnsmasq will check if the 'Server Identifier Override' option was
+set in the incoming packet.
+.IP
+If it was not set, the packet is dropped.
+.IP
+If it was set, dnsmasq evaluates the Server ID against the value provided in \'Server ID Override\' suboption 
+11, as outlined in RFC5107.
+.IP
+In both cases, there is no match against the \'backend\' IP address configured on the interface.
+.IP
+The \fB\-\-dhcp\-allowed\-srvids\fP option allows disabling this security
+mechanism for specific address(es). When enabled, the incoming DHCPREQUEST is
+evaluated against the provided value(s) instead of the addresses configured on
+the local interfaces.
 
 .TP
 .B \-s, --domain=<domain>[[,<address range>[,local]]|<interface>]
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -1175,11 +1175,13 @@
   struct pxe_service *pxe_services;
   struct tag_if *tag_if; 
   struct addr_list *override_relays;
+  struct addr_list *allowed_srvids;
   struct dhcp_relay *relay4, *relay6;
   struct delay_config *delay_conf;
   int override;
   int enable_pxe;
   int doing_ra, doing_dhcp6;
+  int allowing_custom_srvids;
   struct dhcp_netid_list *dhcp_ignore, *dhcp_ignore_names, *dhcp_gen_names; 
   struct dhcp_netid_list *force_broadcast, *bootp_dynamic;
   struct hostsfile *dhcp_hosts_file, *dhcp_opts_file;
--- a/src/option.c
+++ b/src/option.c
@@ -186,6 +186,7 @@
 #define LOPT_STALE_CACHE   377
 #define LOPT_NORR          378
 #define LOPT_NO_IDENT      379
+#define LOPT_DHCP_AL_SVID  386
 
 #ifdef HAVE_GETOPT_LONG
 static const struct option opts[] =  
@@ -376,6 +377,7 @@
     { "fast-dns-retry", 2, 0, LOPT_FAST_RETRY },
     { "use-stale-cache", 2, 0 , LOPT_STALE_CACHE },
     { "no-ident", 0, 0, LOPT_NO_IDENT },
+    { "dhcp-allowed-srvids", 1, 0, LOPT_DHCP_AL_SVID },
     { NULL, 0, 0, 0 }
   };
 
@@ -573,6 +575,7 @@
   { LOPT_QUIET_TFTP, OPT_QUIET_TFTP, NULL, gettext_noop("Do not log routine TFTP."), NULL },
   { LOPT_NORR, OPT_NORR, NULL, gettext_noop("Suppress round-robin ordering of DNS records."), NULL },
   { LOPT_NO_IDENT, OPT_NO_IDENT, NULL, gettext_noop("Do not add CHAOS TXT records."), NULL },
+  { LOPT_DHCP_AL_SVID, ARG_DUP, "[=<ipaddr>]...", gettext_noop("Allow these ServerIDs"), NULL },
   { 0, 0, NULL, NULL, NULL }
 }; 
 
@@ -4632,6 +4635,18 @@
 	
 	break;
       }
+    case LOPT_DHCP_AL_SVID:
+	daemon->allowing_custom_srvids = 1;
+	while (arg) {
+	struct addr_list *new = opt_malloc(sizeof(struct addr_list));
+	comma = split(arg);
+	if (!(inet_pton(AF_INET, arg, &new->addr) > 0))
+	  ret_err_free(_("bad dhcp-allowed-srvids address"), new);
+	new->next = daemon->allowed_srvids;
+	daemon->allowed_srvids = new;
+	arg = comma;
+	}
+	break;
 
 #endif
       
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -1202,8 +1202,22 @@
 	      
 	      if (override.s_addr != 0)
 		{
-		  if (option_addr(opt).s_addr != override.s_addr)
-		    return 0;
+		  if (option_addr(opt).s_addr != override.s_addr) {
+		    if (daemon->allowing_custom_srvids) {
+			struct addr_list *l;
+			for (l = daemon->allowed_srvids; l; l = l->next)
+			  if (l->addr.s_addr == option_addr(opt).s_addr) {
+			    inet_ntop(AF_INET, &l->addr.s_addr, daemon->addrbuff, ADDRSTRLEN);
+			    my_syslog(MS_DHCP | LOG_DEBUG, _("ServerID %s is explicitly allowed."),
+				        daemon->addrbuff);
+			    break;
+			  }
+			if (!l)
+			  return 0;
+		    }
+		   else
+			return 0;
+		  }
 		}
 	      else 
 		{
@@ -1230,12 +1244,28 @@
 			override = intr->addr.in.sin_addr;
 		      else
 			{
-			  /* In auth mode, a REQUEST sent to the wrong server
-			     should be faulted, so that the client establishes 
-			     communication with us, otherwise, silently ignore. */
-			  if (!option_bool(OPT_AUTHORITATIVE))
-			    return 0;
-			  message = _("wrong server-ID");
+			  if (daemon->allowing_custom_srvids) {
+				my_syslog(MS_DHCP | LOG_DEBUG, _("checking allowed custom serverids"));
+				struct addr_list *l;
+				for (l = daemon->allowed_srvids; l; l = l->next)
+				  if (l->addr.s_addr == option_addr(opt).s_addr)
+				    break;
+				if (l) {
+		              inet_ntop(AF_INET, &l->addr.s_addr, daemon->addrbuff, ADDRSTRLEN);
+				  my_syslog(MS_DHCP | LOG_DEBUG, _("ServerID %s is explicitly allowed."),
+						daemon->addrbuff);
+				  override = option_addr(opt);
+				}
+				else
+				{
+				  /* In auth mode, a REQUEST sent to the wrong server
+				  should be faulted, so that the client establishes
+				  communication with us, otherwise, silently ignore. */
+				  if (!option_bool(OPT_AUTHORITATIVE))
+				    return 0;
+				  message = _("wrong server-ID");
+				}
+			  }
 			}
 		    }
 		}
